### PR TITLE
[ADVAPP-811] Introduce the ability to convert a Prospect into a Student

### PR DESCRIPTION
**Pull request by ankit-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-811

### Technical Description

> Introduce a Convert button on a Prospect Record, which covert a prospect into student record.

### Any deployment steps required?

> No

### Are any Feature Flags Added?

> Yes
> Added "convert_prospect_to_student" feature flag.

_______________________________________________

### CODE REVIEW

File: app-modules/prospect/src/Filament/Resources/ProspectResource/Widgets/ConvertedStudentBadge.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+class ConvertedStudentBadge extends BaseWidget
+{
+    public $student;
+
+    protected static bool $isLazy = false;
+
+    protected int | string | array $columnSpan = 'full';
+
+    protected static string $view = 'filament.widgets.student-converted-badge';
```
Comment (Orrison) [inline app-modules/prospect/src/Filament/Resources/ProspectResource/Widgets/ConvertedStudentBadge.php:49]: This view should be put within the `prospects` module and then this reference namespaced

File: app-modules/prospect/tests/Prospect/EditProspectTest.php [hunk 2]
```diff
@@ -149,3 +152,92 @@
...
+    $prospect = Prospect::factory()
+        ->create();
+
+    $student = Student::factory()
+        ->create();
+
+    livewire(EditProspect::class, [
+        'record' => $prospect->getRouteKey(),
+    ])
+        ->mountAction(ConvertToStudent::class)
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/EditProspectTest.php:211]: Is this `mountAction` call needed?

File: app-modules/prospect/tests/Prospect/EditProspectTest.php [hunk 3]
```diff
@@ -149,3 +152,92 @@
...
+        ->callAction(
+            ConvertToStudent::class,
+            data: ['student_id' => $student->getKey()]
+        )->assertSuccessful();
+
+    $prospect->refresh();
+    expect($prospect->student()->exists())->toBeTrue();
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/EditProspectTest.php:218]: You should assert that the relationship is to the `Student` that you expect it to be

File: app-modules/prospect/tests/Prospect/EditProspectTest.php [hunk 4]
```diff
@@ -149,3 +152,92 @@
...
+    actingAs($user);
+
+    $prospect = Prospect::factory()
+        ->for(Student::factory(), 'student')
+        ->create();
+
+    livewire(EditProspect::class, [
+        'record' => $prospect->getRouteKey(),
+    ])
+        ->mountAction(DisassociateStudent::class)
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/EditProspectTest.php:236]: Is this `mountAction` call needed?

File: app-modules/prospect/tests/Prospect/ViewProspectTest.php [hunk 5]
```diff
@@ -68,3 +73,94 @@
...
+    ])
+        ->mountAction(ConvertToStudent::class)
+        ->callAction(
+            ConvertToStudent::class,
+            data: ['student_id' => $student->getKey()]
+        )->assertSuccessful();
+
+    $prospect->refresh();
+
+    expect($prospect->student()->exists())->toBeTrue();
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/ViewProspectTest.php:140]: You should assert that the relationship is to the Student that you would expect.

File: app-modules/prospect/tests/Prospect/ViewProspectTest.php [hunk 6]
```diff
@@ -68,3 +73,94 @@
...
+    $prospect = Prospect::factory()
+        ->create();
+
+    $student = Student::factory()
+        ->create();
+
+    livewire(ViewProspect::class, [
+        'record' => $prospect->getRouteKey(),
+    ])
+        ->mountAction(ConvertToStudent::class)
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/ViewProspectTest.php:132]: Is this `mountAction` call needed?

File: app-modules/prospect/tests/Prospect/ViewProspectTest.php [hunk 7]
```diff
@@ -68,3 +73,94 @@
...
+    actingAs($user);
+
+    $prospect = Prospect::factory()
+        ->for(Student::factory(), 'student')
+        ->create();
+
+    livewire(ViewProspect::class, [
+        'record' => $prospect->getRouteKey(),
+    ])
+        ->mountAction(DisassociateStudent::class)
```
Comment (Orrison) [inline app-modules/prospect/tests/Prospect/ViewProspectTest.php:158]: Is this `mountAction` call needed?

**Review by Orrison — CHANGES_REQUESTED** [review]
This looks good but I had a few notes.

Also overarching, I think it would be best if the placement of the convert notice was changed. I believe it would be best going all across the top of the content. Perhaps the Render Hook `PanelsRenderHook::PAGE_START` may work better than this widget.
Also this notice needs to be present on all pages within the Student Holistic view. Not just View and Edit, but "Emails & Texts" etc.

**Comment by Orrison** [comment]
> > This looks good but I had a few notes.
> > Also overarching, I think it would be best if the placement of the convert notice was changed. I believe it would be best going all across the top of the content. Perhaps the Render Hook `PanelsRenderHook::PAGE_START` may work better than this widget. Also this notice needs to be present on all pages within the Student Holistic view. Not just View and Edit, but "Emails & Texts" etc.
> 
> Hi @Orrison ,
> 
> You mean the Prospect Holistic view here?
> 
> Thank you.

Yes, sorry. I meant the Prospect Holistic View

**Comment by ankit-canyon** [comment]
> > > This looks good but I had a few notes.
> > > Also overarching, I think it would be best if the placement of the convert notice was changed. I believe it would be best going all across the top of the content. Perhaps the Render Hook `PanelsRenderHook::PAGE_START` may work better than this widget. Also this notice needs to be present on all pages within the Student Holistic view. Not just View and Edit, but "Emails & Texts" etc.
> > 
> > 
> > Hi @Orrison ,
> > You mean the Prospect Holistic view here?
> > Thank you.
> 
> Yes, sorry. I meant the Prospect Holistic View

Yep ! I that's I already done with the recent changes.

File: app-modules/prospect/resources/views/student-converted-badge.blade.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    - Use of this software implies agreement to the license terms and conditions as stated
+      in the Elastic License 2.0.
+
+    For more information or inquiries please visit our website at
+    https://www.canyongbs.com or contact us via email at legal@canyongbs.com.
+
+</COPYRIGHT>
+--}}
+@use('AdvisingApp\StudentDataModel\Filament\Resources\StudentResource')
+@if (Feature::active('convert_prospect_to_student') && $this->getRecord()->student)
```
Comment (Orrison) [inline app-modules/prospect/resources/views/student-converted-badge.blade.php:35]: Let's make sure to import the `Feature` class, it's interesting that it seems to be finding it on it's own now. But best to be explicit.

**Review by Orrison — CHANGES_REQUESTED** [review]
Looking good, just a comment.

Also lets make tests ensuring that the blade render hook shows or does not show the banner when the Student is in the correct configuration for each page.

It also seems that the banner is not on the Programs page

**Review by Orrison — CHANGES_REQUESTED** [review]

**Review by Orrison — APPROVED** [review]

**Pull request merged**